superset/security.py

```diff
@@ -77,177 +78,303 @@
+
+        logging.info(
+            'Fetching a set of all perms to lookup which ones are missing')
+        all_pvs = set()
+        for pv in self.get_session.query(self.permissionview_model).all():
+            if pv.permission and pv.view_menu:
+                all_pvs.add((pv.permission.name, pv.view_menu.name))
+
+        def merge_pv(view_menu, perm):
+            """Create permission view menu only if it doesn't exist"""
+            if view_menu and perm and (view_menu, perm) not in all_pvs:
+                self.merge_perm(view_menu, perm)
+
+        logging.info('Creating missing datasource permissions.')
+        datasources = ConnectorRegistry.get_all_datasources(db.session)
+        for datasource in datasources:
+            merge_pv('datasource_access', datasource.get_perm())
+            merge_pv('schema_access', datasource.schema_perm)
+
+        logging.info('Creating missing database permissions.')
+        databases = db.session.query(models.Database).all()
+        for database in databases:
+            merge_pv('database_access', database.perm)
+
+        logging.info('Creating missing metrics permissions')
+        metrics = []
+        for datasource_class in ConnectorRegistry.sources.values():
+            metrics += list(db.session.query(datasource_class.metric_class).all())
+
+        for metric in metrics:
+            if metric.is_restricted:
+                merge_pv('metric_access', metric.perm)
+
+    def clean_perms(self):
+        """FAB leaves faulty permissions that need to be cleaned up"""
+        logging.info('Cleaning faulty perms')
+        sesh = self.get_session
+        pvms = (
+            sesh.query(ab_models.PermissionView)
+            .filter(or_(
+                ab_models.PermissionView.permission == None,  # NOQA
+                ab_models.PermissionView.view_menu == None,  # NOQA
+            ))
+        )
+        deleted_count = pvms.delete()
+        sesh.commit()
+        if deleted_count:
+            logging.info('Deleted {} faulty permissions'.format(deleted_count))
+
+    def sync_role_definitions(self):
+        """Inits the Superset application with security roles and such"""
+        from superset import conf
+        logging.info('Syncing role definition')
+
+        self.get_or_create_main_db()
+        self.create_custom_permissions()
+
+        # Creating default roles
+        self.set_role('Admin', self.is_admin_pvm)
+        self.set_role('Alpha', self.is_alpha_pvm)
+        self.set_role('Gamma', self.is_gamma_pvm)
+        self.set_role('granter', self.is_granter_pvm)
+        self.set_role('sql_lab', self.is_sql_lab_pvm)
+
+        if conf.get('PUBLIC_ROLE_LIKE_GAMMA', False):
+            self.set_role('Public', self.is_gamma_pvm)
+
+        self.create_missing_perms()
+
+        # commit role and view menu updates
+        self.get_session.commit()
+        self.clean_perms()
+
+    def get_or_create_main_db(self):
+        from superset import conf, db
```

Comment:
I moved them there to avoid cyclical dependencies. Like you mentioned above, get_or_create_main_db seems oddly placed in this file so its not surprising that it has multiple lines of import. I will move it elsewhere in a follow-up PR.